src/modules/dashboard/components/Dialogs/AddExistingSafeDialog/AddExistingSafeDialog.tsx

```diff
@@ -1,53 +1,278 @@
+    defaultMessage: 'Please enter a safe name',
+  },
+  moduleAddressError: {
+    id: 'dashboard.AddExistingSafeDialog.moduleAddressError',
+    defaultMessage: 'Please enter a module address',
+  },
+  safeAlreadyExistsError: {
+    id: 'dashboard.AddExistingSafeDialog.safeAlreadyExistsError',
+    defaultMessage: 'Safe already exists in this colony.',
+  },
+  fetchFailedError: {
+    id: 'dashboard.AddExistingSafeDialog.fetchFailedError',
+    defaultMessage: `Could not fetch {type} details. Please check your connection and try again.`,
+  },
+  safeNotFoundError: {
+    id: 'dashboard.AddExistingSafeDialog.safeNotFoundError',
+    defaultMessage: `Safe not found on {selectedChain}`,
+  },
+  moduleNotConnectedError: {
+    id: 'dashboard.AddExistingSafeDialog.moduleNotConnectedError',
+    defaultMessage: `Module not connected to safe on {selectedChain}`,
+  },
+  moduleNotFoundError: {
+    id: 'dashboard.AddExistingSafeDialog.moduleNotFoundError',
+    defaultMessage: `Module not found on {selectedChain}`,
+  },
+});
 const displayName = 'dashboard.AddExistingSafeDialog';
 
+// "this" is needed within yup.test().
+/* eslint-disable react/no-this-in-sfc */
 const AddExistingSafeDialog = ({
   colony: { colonyAddress, colonyName, safes },
   callStep,
   prevStep,
   cancel,
   close,
 }: Props) => {
+  const { formatMessage } = useIntl();
   const history = useHistory();
+  const loadingSafeState = useState<boolean>(false);
+  const loadingModuleState = useState<boolean>(false);
+  const [stepIndex, setStepIndex] = useState<number>(1);
+  const [abortController, setAbortController] = useState<
+    AbortController | undefined
+  >(undefined);
+
+  function getFetchErrorMsg(
+    e: Error,
+    type: 'Safe' | 'Module',
+  ): yup.ValidationError {
+    // If fetch was aborted
+    if (e instanceof DOMException && e.message.includes('abort')) {
+      return this.createError({
+        message: FETCH_ABORTED,
+      });
+    }
+
+    // If fetching produces an error (e.g. network error)
+    return this.createError({
+      message: formatMessage(MSG.fetchFailedError, { type }),
+    });
+  }
+  const handleTestCompletion = (
+    result: true | yup.ValidationError,
+    type: 'Safe' | 'Module',
+  ): Promise<true | yup.ValidationError> => {
+    const isFetchAborted = result !== true && result.message === FETCH_ABORTED;
+    const loadingState =
+      type === 'Safe' ? loadingSafeState : loadingModuleState;
+    const [, setIsLoadingState] = loadingState;
+
+    if (!isFetchAborted) {
+      setIsLoadingState(false);
+      return new Promise((res) => res(result));
+    }
+    // If fetching was aborted, don't show error.
+    return new Promise((res) => res(true));
+  };
   const validationSchema = yup.object().shape({
     chainId: yup.string().required(),
-    contractAddress: yup.string().address().required(),
-    safeName: yup.string().required().max(20),
+    contractAddress: yup
```

Comment:
What do you think of taking validationSchema out to split the code a bit? The component seems quite loaded...

Reply:
@chinins I think this is a very good idea and have implemented.